WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,178 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	/// <seealso href="https://murch.one/wp-content/uploads/2016/11/erhardt2016coinselection.pdf">Section "5.3 Branch and Bound".</seealso>
+	public class BranchAndBound
+	{
+		private readonly Random _random = new();
+
+		public BranchAndBound(List<long> values)
+		{
+			Count = values.Count;
+			SortedValues = values.OrderByDescending(x => x).ToArray();
+		}
+
+		private enum NextAction
+		{
+			/// <summary>First try to include a value and then try not to include the value in the selection.</summary>
+			AandB,
+
+			/// <summary>First try NOT to include a value and then try to include the value in the selection.</summary>
+			BandA,
+
+			/// <summary>Include value.</summary>
+			A,
+
+			/// <summary>Omit value.</summary>
+			B,
+
+			/// <summary>Current selection is wrong, rolling back and trying different combination.</summary>
+			Backtrack
+		}
+
+		/// <remarks>Input values sorted in descending order.</remarks>
+		private long[] SortedValues { get; }
+
+		/// <summary>Number of input values.</summary>
+		private int Count { get; }
+
+		/// <summary>
+		/// Attempts to find a set of values that sum up to the target value.
+		/// </summary>
+		/// <param name="target">Target value we want to sum up from the input values.</param>
+		/// <param name="selectedValues">Values that sum up to the <paramref name="target"/> value.</param>
+		/// <returns><c>true</c> when a match is found, <c>false</c> otherwise.</returns>
+		public bool TryGetExactMatch(long target, [NotNullWhen(true)] out List<long>? selectedValues)
+		{
+			using (BenchmarkLogger.Measure())
```

Comment:
This cross cutting concepts shouldn't be part of the code. We can measure the times using special tools when needed or implemented using a decorator when needed too.

Reply:
Yes, removed.